From: Stained Glass OS <ke7oxh@gmail.com>
Subject: [PATCH] msvcp90: Implement _Xfsopen and export tr1::_Xinvalid.

_Xfsopen in msvcp80 and msvcp90, the wide-name open of the file buffers
(_wfsopen), and msvcp90's tr1::_Xinvalid, which throws
std::invalid_argument with the given message like _Xinvalid_argument.
---
 dlls/msvcp80/msvcp80.spec |  4 ++--
 dlls/msvcp90/ios.c        | 14 ++++++++++++++
 dlls/msvcp90/msvcp90.spec |  8 ++++----
 3 files changed, 20 insertions(+), 6 deletions(-)

diff --git a/dlls/msvcp80/msvcp80.spec b/dlls/msvcp80/msvcp80.spec
index 726ff93..cf207dd 100644
--- a/dlls/msvcp80/msvcp80.spec
+++ b/dlls/msvcp80/msvcp80.spec
@@ -3769,8 +3769,8 @@
 @ cdecl -arch=arm ?_Widen_s@?$ctype@_W@std@@QBAPBDPBD0PA_WI@Z(ptr ptr ptr ptr long) ctype_wchar__Widen_s
 @ thiscall -arch=i386 ?_Widen_s@?$ctype@_W@std@@QBEPBDPBD0PA_WI@Z(ptr ptr ptr ptr long) ctype_wchar__Widen_s
 @ cdecl -arch=win64 ?_Widen_s@?$ctype@_W@std@@QEBAPEBDPEBD0PEA_W_K@Z(ptr ptr ptr ptr long) ctype_wchar__Widen_s
-@ stub -arch=win32 ?_Xfsopen@std@@YAPAU_iobuf@@PB_W0H@Z
-@ stub -arch=win64 ?_Xfsopen@std@@YAPEAU_iobuf@@PEB_W0H@Z
+@ cdecl -arch=win32 ?_Xfsopen@std@@YAPAU_iobuf@@PB_W0H@Z(wstr wstr long) _Xfsopen
+@ cdecl -arch=win64 ?_Xfsopen@std@@YAPEAU_iobuf@@PEB_W0H@Z(wstr wstr long) _Xfsopen
 @ cdecl ?_Xinvarg@_String_base@std@@SAXXZ() MSVCP__String_base_Xinvarg
 @ cdecl ?_Xlen@_String_base@std@@SAXXZ() MSVCP__String_base_Xlen
 @ cdecl ?_Xran@_String_base@std@@SAXXZ() MSVCP__String_base_Xran
diff --git a/dlls/msvcp90/ios.c b/dlls/msvcp90/ios.c
index 28bc307..69fae06 100644
--- a/dlls/msvcp90/ios.c
+++ b/dlls/msvcp90/ios.c
@@ -3316,6 +3316,20 @@ bool __thiscall basic_filebuf_char_is_open(const basic_filebuf_char *this)
     return this->file != NULL;
 }
 
+#if _MSVCP_VER >= 80 && _MSVCP_VER <= 90
+/* ?_Xfsopen@std@@YAPAU_iobuf@@PB_W0H@Z */
+/* ?_Xfsopen@std@@YAPEAU_iobuf@@PEB_W0H@Z */
+/* the wide-name open of the file buffers: _wfsopen */
+FILE* __cdecl _Xfsopen(const wchar_t *name, const wchar_t *mode, int prot)
+{
+    TRACE("(%s %s %d)\n", debugstr_w(name), debugstr_w(mode), prot);
+#ifdef SG_MUTANT_XFSOPEN_NO_MODE
+    return _wfsopen(name, L"r", prot);
+#endif
+    return _wfsopen(name, mode, prot);
+}
+#endif
+
 /* ?_Fiopen@std@@YAPAU_iobuf@@PB_WHH@Z */
 /* ?_Fiopen@std@@YAPEAU_iobuf@@PEB_WHH@Z */
 FILE* __cdecl _Fiopen_wchar(const wchar_t *name, int mode, int prot)
diff --git a/dlls/msvcp90/msvcp90.spec b/dlls/msvcp90/msvcp90.spec
index cccfac4..7298f7c 100644
--- a/dlls/msvcp90/msvcp90.spec
+++ b/dlls/msvcp90/msvcp90.spec
@@ -3803,11 +3803,11 @@
 @ cdecl ?_XLgamma@tr1@std@@YANN@Z(double) std__XLgamma_double
 @ cdecl ?_XLgamma@tr1@std@@YAOO@Z(double) std__XLgamma_double
 @ stub ?_Xbad@tr1@std@@YAXW4error_type@regex_constants@12@@Z
-@ stub -arch=win32 ?_Xfsopen@std@@YAPAU_iobuf@@PB_W0H@Z
-@ stub -arch=win64 ?_Xfsopen@std@@YAPEAU_iobuf@@PEB_W0H@Z
+@ cdecl -arch=win32 ?_Xfsopen@std@@YAPAU_iobuf@@PB_W0H@Z(wstr wstr long) _Xfsopen
+@ cdecl -arch=win64 ?_Xfsopen@std@@YAPEAU_iobuf@@PEB_W0H@Z(wstr wstr long) _Xfsopen
 @ stub ?_Xfunc@tr1@std@@YAXXZ
-@ stub -arch=win32 ?_Xinvalid@tr1@std@@YAXPBD@Z
-@ stub -arch=win64 ?_Xinvalid@tr1@std@@YAXPEBD@Z
+@ cdecl -arch=win32 ?_Xinvalid@tr1@std@@YAXPBD@Z(str) _Xinvalid_argument
+@ cdecl -arch=win64 ?_Xinvalid@tr1@std@@YAXPEBD@Z(str) _Xinvalid_argument
 @ cdecl ?_Xinvarg@_String_base@std@@SAXXZ() MSVCP__String_base_Xinvarg
 @ cdecl ?_Xlen@_String_base@std@@SAXXZ() MSVCP__String_base_Xlen
 @ cdecl ?_Xmem@tr1@std@@YAXXZ() _Xmem
